tmt/cli.py

```diff
@@ -2081,3 +2081,100 @@ def completion_fish(context: Context, install: bool, **kwargs: Any) -> None:
     Setup shell completions for fish.
     """
     setup_completion('fish', install)
+
+
+@click.command(context_settings={
+    "ignore_unknown_options": True,
+    "allow_extra_args": True,
+    })
+@click.pass_context
+@click.option(
+    '-r', '--root', metavar='PATH', show_default=True, default='.',
+    help='Variable passed to tmt main. See `tmt --help`')
+@click.option(
+    '-v', '--verbose', count=True, default=0,
+    help='Variable passed to tmt main. See `tmt --help`')
+@click.option(
+    '-d', '--debug', count=True, default=0,
+    help='Variable passed to tmt main. See `tmt --help`')
```

Comment:
Both `debug` and `verbose` should be added via `@verbosity_options`, is there a reason against using it? I suppose that the user does not need to know the value is passed to `tmt main` as long as it has the expected effect, therefore the help message can remain the same.

I would point out the technical focus in the case of other options too - as a mere user, I have no idea what "tmt main" might mean. What I would appreciate in a help text would be what the option does - the fact it's consumed by the function below and passed to `main()` is not as important to me as the effect itself. For that purpose, I would recommend re-using the help text from the original options, i.e. "Show more details. Use multiple times to raise verbosity." instead of "Variable passed to tmt main. See `tmt --help`". Will `tmt-pre-commit` be verbose when I use `-v`? If so, I don't care about how it's done internally :) We should probably add reusable decorators for `--version` and `--root`, to avoid copying & pasting.

Reply:
Some context, the main thing why the different entry-point was written is because in the click cli is that the `--root` option is an option for top-level `tmt`. I probably tried to re-define `--root` at the `tmt lint` level but wasn't able to, don't remember that context. Technically if `tmt lint` could forward the `--root` it is the preferred approach.

But this pre-commit entry-point is not meant to be used in any context outside of `pre-commit-hooks.yaml`, therefore it shouldn't support the full `tmt` cli. E.g. the user would not be using `--version` in `pre-commit-config.yaml`.

`verbose` and `debug` are a similar story. If pre-commit offers an interface to inject verbose flag, than yes it should support it, otherwise the user should use `tmt lint` to check manually with these flags. And there's nuance if pre-commit does not support multiple verbose flags.

Reply:
Why I've added all of those options when I wrote the PR 🤷‍♂️. Hindsight is 20/20. Probably better to discuss what other alternative designs for this problem are there. I don't like having a separate executable just for pre-commit, but we can't use modues directly can we?

Another possibility would be to have a non-click entry-point that simply filters out `--root` and passes everything else to `main()`, but I thought I've had in that form initially? Maybe I was thinking click can do some magical handling. Could re-write that to be just 3-4 lines of code.

Reply:
> But this pre-commit entry-point is not meant to be used in any context outside of `pre-commit-hooks.yaml`,

Understood, yes.

> therefore it shouldn't support the full `tmt` cli.

It does not have to, and it already diverged and has an option not accepted by the main `tmt` command :) That I'd say is perfectly fine.

> E.g. the user would not be using `--version` in `pre-commit-config.yaml`.

On the other hand, this one might be handy if supported. I would not expect it to be the default part of `.pre-commit-config.yaml` section for tmt lint, but I can easily imagine a user debugging issues with pre-commit might use it in an ad hoc manner to make sure pre-commit is running the right `tmt-lint`.

> `verbose` and `debug` are a similar story. If pre-commit offers an interface to inject verbose flag, than yes it should support it

Depends. I'm not sure whether `pre-commit -vvvvv` would manifest as `tmt-lint -vvvvv` - most likely it would not. But that's just one case, and there's still the `.pre-commit-config.yaml` where I could increase the verbosity of `tmt-lint` check for all runs, e.g. `args: ['-vvvv']. That is definitely supported and often used, I saw many projects running e.g. `mypy -v`.

> otherwise the user should use `tmt lint` to check manually with these flags.

They still can do it. But the environment in pre-commit stage might differ, there may be different results reported by a check inside and outside of pre-commit (mypy does exactly that - reports no issues in pre-commit, reports several troubling lines when executed under `hatch`...). I wouldn't rob users of a chance to increase verbosity through the config file.

Reply:
Looking at the [`pre-commit-hooks.yaml`](https://pre-commit.com/#creating-new-hooks) and `pre-commit run --help`, I do not see any indication that we can pass different `args` then those in `pre-commit-config.yaml`. I did not expect editing `args` as part of debug workflow because I am often too paranoid that I will forget to remove that before commit. But I can see that being useful for `mypy` debugging since the `pre-commit` cached `venv` can often be out-of-sync with the local environment.

> mypy does exactly that - reports no issues in pre-commit

Btw `pre-commit` authors have strong opinions that `additional-dependencies` should be pinned so that the cache key changes when dependencies change. Personally I think that is a maintenance nightmare and it makes more sense to drop the `mypy/playwright` pre-commits in favor of https://github.com/teemtee/tmt/pull/2841 and `hatch` command where there is more control on the environment.